contracts/src/SimplePDPService.sol

```diff
@@ -69,50 +128,51 @@ require(_pdpServiceAddress != address(0), "PDP service address cannot be zero");
         receiveProofSetEvent(proofSetId, OperationType.PROVE_POSSESSION, abi.encode(challengedLeafCount, seed, challengeCount));
+        emit Debug("Here we go", 0);
+        if (provenThisPeriod[proofSetId]) { 
+            // return immediately, we've already witnessed a proof for this proof set this period
+            return; 
+        }
+
+        if (challengeCount < getChallengesPerProof()) {
+            revert("Invalid challenge count < 5");
+        }
+        emit Debug("deadline", provingDeadlines[proofSetId]);
+        emit Debug("block number", block.number);
+        // check for late proof 
+        if (provingDeadlines[proofSetId] < block.number) {
+            uint256 periodsLate = 1 + ((block.number - provingDeadlines[proofSetId]) / getMaxProvingPeriod());
+            emit Debug("we're late", periodsLate); 
+            emit FaultRecord(FaultType.LATE, periodsLate);
+        }
+        provenThisPeriod[proofSetId] = true;
     }
 
+    // nextProvingPeriod checks for unsubmitted proof and emits a fault record if so
     function nextProvingPeriod(uint256 proofSetId, uint256 leafCount) external onlyPDPVerifier {
         receiveProofSetEvent(proofSetId, OperationType.NEXT_PROVING_PERIOD, abi.encode(leafCount));
-    }
-
-    function receiveProofSetEvent(uint256 proofSetId, OperationType operationType, bytes memory extraData ) internal {
-        uint64 epoch = uint64(block.number);
-        EventRecord memory newRecord = EventRecord({
-            epoch: epoch,
-            proofSetId: proofSetId,
-            operationType: operationType,
-            extraData: extraData
-        });
-        proofSetEvents[proofSetId].push(newRecord);
-        emit RecordAdded(proofSetId, epoch, operationType);
-    }
-
-    // Function to get the number of events for a proof set
-    function getEventCount(uint256 proofSetId) external view returns (uint256) {
-        return proofSetEvents[proofSetId].length;
-    }
-
-    // Function to get a specific event for a proof set
-    function getEvent(uint256 proofSetId, uint256 eventIndex)
-        external
-        view
-        returns (EventRecord memory)
-    {
-        require(eventIndex < proofSetEvents[proofSetId].length, "Event index out of bounds");
-        return proofSetEvents[proofSetId][eventIndex];
-    }
-
-    // Function to get all events for a proof set
-    function listEvents(uint256 proofSetId) external view returns (EventRecord[] memory) {
-        return proofSetEvents[proofSetId];
+        // check for unsubmitted proof 
+        if (!provenThisPeriod[proofSetId]) {
+            uint256 periodsSkipped = 1;
+            if (provingDeadlines[proofSetId] < block.number) {
+                periodsSkipped = 1 + ((block.number - provingDeadlines[proofSetId]) / getMaxProvingPeriod());
+            }
+            emit FaultRecord(FaultType.SKIPPED, periodsSkipped);
+        }
+        provenThisPeriod[proofSetId] = false;
```

Comment:
This appears to be missing re-setting `provingDeadlines[proofSetId]` to a future epoch.

There's a policy decision to be made about which future epoch. I reckon that if it _was_ proven this period, we should add the full proving period to the deadline (so it stays the same time of day). So proving early doesn't lead to disadvantage of needing to prove more frequently over the long term.

But if this period wasn't proven, we probably want to set the deadline based on current epoch, like when the first root is added.